### DISCUSSION TITLE
How to build m5 (Apple aarch64)

### DISCUSSION THREAD

**Discussion by MarcoBonino** [gem5/gem5#1728] [Q&A] [answered]
Hello!

I'm trying to build m5 tool using Apple M3.
I'm trying directly on the machine and using ubuntu 24 on a virtual machine.

If I understood correctly we should build m5 for the host architecture: it is a normal binary running on the machine we are using.

Here how I build: `scons build/arm64/out/m5`
I always get the following error: `Error: No config found, and no implicit config recognized`

Could you help on build it ?

Here full log:
```
scons: Reading SConscript files ...
Mkdir("/Users/makko/Projects/gem5/build/arm64/gem5.build")
Checking for linker -Wl,--as-needed support... no
Checking for compiler -Wno-c99-designator support... yes
Checking for compiler -Wno-defaulted-function-deleted support... yes
Checking for compiler -gz support... yes
Checking for linker -gz support... yes
Info: Using Python config: python3-config
Checking for C header file Python.h... yes
Checking Python version... 3.13.0
Checking for accept(0,0,0) in C++ library None... yes
Checking for zlibVersion() in C++ library z... yes
Checking for C library tcmalloc_minimal... no
Checking for C library tcmalloc... no
Warning: You can get a 12% performance improvement by installing tcmalloc (libgoogle-perftools-dev package on Ubuntu or RedHat).
Building in /Users/makko/Projects/gem5/build/arm64
"build_tools/kconfig_base.py" "/Users/makko/Projects/gem5/build/arm64/gem5.build/Kconfig" "/Users/makko/Projects/gem5/src/Kconfig"
Warning: While checking protoc version: [Errno 2] No such file or directory: 'protoc'
Warning: Protocol buffer compiler (protoc) not found.
         Please install protobuf-compiler for tracing support.
Checking for C header file capstone/capstone.h... no
Warning: Header file <capstone/capstone.h> not found.
         This host has no capstone library installed.
Checking for C header file linux/kvm.h... no
Warning: Info: Compatible header file <linux/kvm.h> not found, disabling KVM support.
Checking for C header file linux/if_tun.h... no
Info: Compatible header file <linux/if_tun.h> not found.
Checking for backtrace_symbols_fd((void *)1, 0, 0) in C library None... yes
Checking for C header file fenv.h... yes
Checking for C header file png.h... no
Warning: Header file <png.h> not found.
         This host has no libpng library.
         Disabling support for PNG framebuffers.
Checking for clock_nanosleep(0,0,NULL,NULL) in C library None... no
Checking for clock_nanosleep(0,0,NULL,NULL) in C library rt... no
Warning: Can't find library for POSIX clocks.
Checking for C header file valgrind/valgrind.h... no
Checking for H5Fcreate("", 0, 0, 0) in C library hdf5... no
Warning: Couldn't find HDF5 C++ libraries. Disabling HDF5 support.
Checking for shm_open("/test", 0, 0) in C library None... yes
Warning: Cannot enable KVM, host seems to lack KVM support
Error: No config found, and no implicit config recognized
```

**Comment by BobbyRBruce** [comment]
I think you're building from the wrong directory.  You need to run `scons build/arm64/out/m5` from the "util/m5" directory

**Reply by MarcoBonino** [reply]
Thanks for pointing out this! That was the reason.
Now I'm getting an error since I do not have the arm64 cross compiler installed:
```
scons build/arm64/out/m5 --verbose                                                                                                                              2 ↵
scons: Reading SConscript files ...
Checking for java package org.junit...(cached) no
junit test framework not found, not build java wrapper test
Checking whether pkg-config program exists...no
pkg-config not detected, can't check for lua51.
Checking for pkg-config package lua51...(cached) no
lua 5.1 not detected, not building lua wrapper.
scons: done reading SConscript files.
scons: Building targets ...
aarch64-linux-gnu-g++ -o build/arm64/call_type/inst.o -c -O2 -std=c++14 -DM5OP_ADDR=0x10010000 -DCALL_TYPE_IS_DEFAULT=true -O2 -DM5OP_ADDR=0x10010000 -I/Users/makko/Projects/gem5/include -I/Users/makko/Projects/gem5/ext/googletest/googletest/include -I/Users/makko/Projects/gem5/ext/googletest/googlemock/include -Ibuild/arm64 build/arm64/call_type/inst.cc
sh: aarch64-linux-gnu-g++: command not found
scons: *** [build/arm64/call_type/inst.o] Error 127
scons: building terminated because of errors.
```

What is not fully clear to me is what architecture we should specify after build/ ? The target or the host ?
My setup is from an Apple M3 (arm64) and I want to run simulations for riscv architecture.

I thought m5 is only used to interact with the gem5 simulator...

**Reply by BobbyRBruce** [reply]
> Thanks for pointing out this! That was the reason. Now I'm getting an error since I do not have the arm64 cross compiler installed:
> 
> ```
> scons build/arm64/out/m5 --verbose                                                                                                                              2 ↵
> scons: Reading SConscript files ...
> Checking for java package org.junit...(cached) no
> junit test framework not found, not build java wrapper test
> Checking whether pkg-config program exists...no
> pkg-config not detected, can't check for lua51.
> Checking for pkg-config package lua51...(cached) no
> lua 5.1 not detected, not building lua wrapper.
> scons: done reading SConscript files.
> scons: Building targets ...
> aarch64-linux-gnu-g++ -o build/arm64/call_type/inst.o -c -O2 -std=c++14 -DM5OP_ADDR=0x10010000 -DCALL_TYPE_IS_DEFAULT=true -O2 -DM5OP_ADDR=0x10010000 -I/Users/makko/Projects/gem5/include -I/Users/makko/Projects/gem5/ext/googletest/googletest/include -I/Users/makko/Projects/gem5/ext/googletest/googlemock/include -Ibuild/arm64 build/arm64/call_type/inst.cc
> sh: aarch64-linux-gnu-g++: command not found
> scons: *** [build/arm64/call_type/inst.o] Error 127
> scons: building terminated because of errors.
> ```
> 
> What is not fully clear to me is what architecture we should specify after build/ ? The target or the host ? My setup is from an Apple M3 (arm64) and I want to run simulations for riscv architecture.
> 
> I thought m5 is only used to interact with the gem5 simulator...

Unfortunately, someone just hardcoded the use of the cross-compilers in these makefiles with the assumption that you're on an X86 machine. The only workaround I have is if you use a Docker container to emulate an X86 host and compile from there. Here's the command that you run from the root of the gem5 repo:

```
docker run --rm -it -v $(pwd):/gem5 -w    /gem5/util/m5  --platform linux/amd64 ghcr.io/gem5/ubuntu-24.04_all-dependencies:latest bash -c 'apt update -y &&  apt install -y g++-aarch64-* && scons arm64.CROSS_COMPILE=aarch64-linux-gnu- build/arm64/out/m5 -j `nproc`
```

If that doesn't work, you may have to hack away at the makefile to get it building.

**Reply by MarcoBonino** [reply]
that seems to work, but it means that we can't run the executable in macos...
Can we create a task to fix this ?

**Reply by BobbyRBruce** [reply]
Well, you're always going to run it in a gem5 simulation, right? So not a problem as far as far as I can see.

**Reply by MarcoBonino** [reply]
Not sure I understand... with the docker command I can build m5 for linux aarch64, but since I'm running gem5 on macos I need also m5 to be built for macos... 
Could you just clarify one point: m5 will run in the guest operating system or the host ? 
I was assuming the tool is used to trigger checkpoints (and other things) by interacting with the simulator (on the host machine).

**Reply by BobbyRBruce** [reply, marked as answer]
I think you've got it a little backwards: you compile binaries liking to this library and run those binaries inside a gem5 simulation. Using functions from the library you can I "communicate" from the simulated guest system to host. The most common use case here would be to use this library to signal within the chest the host to halt the simulation and exit the simulation. This is often to do something (like take a checkpoint)

**Reply by MarcoBonino** [reply]
Thanks for the explanation!

**Reply by german-v-t-um-es** [reply]
Hi, I am trying to build the gem5.opt for the VEGA_X86 simulator on an M1 chip and I keep getting errors while executing scons over the base repository from gem5. I'm using the docker image from gem5 and the cross compiler that @BobbyRBruce told about. Do you have any idea why these errors could be happening? Thanks for the help.

**Reply by studyztp** [reply]
> > Thanks for pointing out this! That was the reason. Now I'm getting an error since I do not have the arm64 cross compiler installed:
> > ```
> > scons build/arm64/out/m5 --verbose                                                                                                                              2 ↵
> > scons: Reading SConscript files ...
> > Checking for java package org.junit...(cached) no
> > junit test framework not found, not build java wrapper test
> > Checking whether pkg-config program exists...no
> > pkg-config not detected, can't check for lua51.
> > Checking for pkg-config package lua51...(cached) no
> > lua 5.1 not detected, not building lua wrapper.
> > scons: done reading SConscript files.
> > scons: Building targets ...
> > aarch64-linux-gnu-g++ -o build/arm64/call_type/inst.o -c -O2 -std=c++14 -DM5OP_ADDR=0x10010000 -DCALL_TYPE_IS_DEFAULT=true -O2 -DM5OP_ADDR=0x10010000 -I/Users/makko/Projects/gem5/include -I/Users/makko/Projects/gem5/ext/googletest/googletest/include -I/Users/makko/Projects/gem5/ext/googletest/googlemock/include -Ibuild/arm64 build/arm64/call_type/inst.cc
> > sh: aarch64-linux-gnu-g++: command not found
> > scons: *** [build/arm64/call_type/inst.o] Error 127
> > scons: building terminated because of errors.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What is not fully clear to me is what architecture we should specify after build/ ? The target or the host ? My setup is from an Apple M3 (arm64) and I want to run simulations for riscv architecture.
> > I thought m5 is only used to interact with the gem5 simulator...
> 
> Unfortunately, someone just hardcoded the use of the cross-compilers in these makefiles with the assumption that you're on an X86 machine. The only workaround I have is if you use a Docker container to emulate an X86 host and compile from there. Here's the command that you run from the root of the gem5 repo:
> 
> ```
> docker run --rm -it -v $(pwd):/gem5 -w    /gem5/util/m5  --platform linux/amd64 ghcr.io/gem5/ubuntu-24.04_all-dependencies:latest bash -c 'apt update -y &&  apt install -y g++-aarch64-* && scons arm64.CROSS_COMPILE=aarch64-linux-gnu- build/arm64/out/m5 -j `nproc`
> ```
> 
> If that doesn't work, you may have to hack away at the makefile to get it building.

The hardcoded part can be overwritten by passing the environment variable `[abi].CROSS_COMPILE=[the one you use]`. For this specific case, it can be overwritten by `arm64.CROSS_COMPILE=`, then it will use the default compilers.

**Reply by myzinsky** [reply]
I tried that what you suggested @studyztp, however, it seems that other issues prevent to use the Darwin toolchain. I can work on linux as well but it would be nice to work natively on my Mac too. Any thoughts?

```bash
❯ scons build/arm64/out/m5 -Q arm64.CROSS_COMPILE=aarch64-apple-darwin24-
junit test framework not found, not build java wrapper test
lua 5.1 not detected, not building lua wrapper.
CC build/arm64/m5_mmap.o
ASPP build/arm64/abi/arm64/m5op.o
ASPP build/arm64/abi/arm64/m5op_addr.o
<instantiation>:4:9: error: ADR/ADRP relocations must be GOT relative
        adrp x9, m5_mem
        ^
<instantiation>:4:9: error: unknown AArch64 fixup kind!
        adrp x9, m5_mem
        ^
<instantiation>:5:9: error: unknown AArch64 fixup kind!
        ldr x9, [ x9, #:lo12:m5_mem ]
        ^
scons: *** [build/arm64/abi/arm64/m5op_addr.o] Error 1
```

**Reply by BobbyRBruce** [reply]
> I tried that what you suggested @studyztp, however, it seems that other issues prevent to use the Darwin toolchain. I can work on linux as well but it would be nice to work natively on my Mac too. Any thoughts?
> 
> ```shell
> ❯ scons build/arm64/out/m5 -Q arm64.CROSS_COMPILE=aarch64-apple-darwin24-
> junit test framework not found, not build java wrapper test
> lua 5.1 not detected, not building lua wrapper.
> CC build/arm64/m5_mmap.o
> ASPP build/arm64/abi/arm64/m5op.o
> ASPP build/arm64/abi/arm64/m5op_addr.o
> <instantiation>:4:9: error: ADR/ADRP relocations must be GOT relative
>         adrp x9, m5_mem
>         ^
> <instantiation>:4:9: error: unknown AArch64 fixup kind!
>         adrp x9, m5_mem
>         ^
> <instantiation>:5:9: error: unknown AArch64 fixup kind!
>         ldr x9, [ x9, #:lo12:m5_mem ]
>         ^
> scons: *** [build/arm64/abi/arm64/m5op_addr.o] Error 1
> ```

The m5 utility is a bit annoying because it compilers some pre-prepared assembly into the binary and that's not as portable as compiling from source on each system.

## Potential solution

((Note: This is just the answer ChatGPT gave me, but it seems like it could work)). 

Dlete the following lines from "util/m5/src/abi/arm64/m5op_addr.S":

```asm
 #else
         // normally.
         adrp x9, m5_mem
         ldr x9, [ x9, #:lo12:m5_mem ]
 #endif
```

## Explanation

This branch its not valid because macOS requires global symbol references like `m5_mem` to be GOT-relative when used in adrp.

**Reply by myzinsky** [reply]
Okay, then I get:

```
ld: library 'crt0.o' not found
```

**Reply by giactra** [reply]
> Dlete the following lines from "util/m5/src/abi/arm64/m5op_addr.S":
> 
>  #else
>          // normally.
>          adrp x9, m5_mem
>          ldr x9, [ x9, #:lo12:m5_mem ]
> 

I feel a better solution would be to try to understand why M5OP_PIC is disregarded in macOS (should be defined by default unless manually disabled)

https://github.com/gem5/gem5/blob/stable/util/m5/src/SConscript#L78

 
 

> Okay, then I get:
> 
> ld: library 'crt0.o' not found

I have seen this several times. Try to locate the object file and symlink it to /usr/lib/rct0.o (that should be on the search path of the dynamic linker IIRC)
